### PR TITLE
dialects: adding Pure() to arith dialect

### PR DESCRIPTION
**Pull request by dshaaban01**
adding Pure to arith dialect so that dce works on arith peephole rewrites

### CODE REVIEW

**Comment by superlopuh** [comment]
Does `make test` pass locally?

File: xdsl/dialects/arith.py [hunk 1]
```diff
@@ -352,11 +351,15 @@ def infer_overflow_type(input_type: Attribute) -> Attribute:
 class Muli(SignlessIntegerBinaryOp):
```
Comment (superlopuh) [inline xdsl/dialects/arith.py:351]: aren't all the `SignlessIntegerBinaryOp`s pure?

**Review by superlopuh — COMMENTED** [review]
Couldn't the abstract superclasses be annotated, instead of the subclasses?

File: xdsl/transforms/constant_fold_interp.py [hunk 2]
```diff
@@ -27,7 +27,7 @@ class ConstantFoldInterpPattern(RewritePattern):
     interpreter: Interpreter
 
     def match_and_rewrite(self, op: Operation, rewriter: PatternRewriter, /):
-        if not op.has_trait(Pure):
```
Comment (webmiche) [inline xdsl/transforms/constant_fold_interp.py:30]: I'd prefer either having Constantlike inherit from Pure or doing this in a different PR.

File: xdsl/dialects/arith.py [hunk 3]
```diff
@@ -767,21 +769,25 @@ def print(self, printer: Printer):
 @irdl_op_definition
 class Addf(FloatingPointLikeBinaryOp):
     name = "arith.addf"
+    traits = frozenset((Pure(),))
```
Comment (webmiche) [inline xdsl/dialects/arith.py:772]: Can we not just add the trait on the parent op? So on FloatingPointLikeBinaryOp in this case.

**Review by webmiche — COMMENTED** [review]
Do you think you can easily add tests for this? This does change some interaction with the framework, right?

**Comment by dshaaban01** [comment]
> Does `make test` pass locally?

Yes it does. Not sure why the tests fail.

**Comment by superlopuh** [comment]
Can you reproduce locally?

**Comment by superlopuh** [comment]
dce only acts on Pure ops, so it looks like it's changing the output of that test

**Comment by superlopuh** [comment]
We'll probably want to add the newly-pure ops to that test to make sure we are consistent in eliminating duplicates with the same inputs

File: xdsl/transforms/constant_fold_interp.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/transforms/constant_fold_interp.py:30]: I agree, could you please split this out into a separate PR?

**Review by superlopuh — COMMENTED** [review]
Could you please add tests for these?

**Comment by superlopuh** [comment]
Let's close this for now?

**Comment by superlopuh** [comment]
Closing for now, let's reopen when we get back to this, there have been some changes both to arith and the impact of Pure with the effects interfaces, it'll be worth revisiting the PR anyway
